Source/WebCore/page/SecurityOriginData.h

```diff
@@ -25,36 +25,52 @@
 
 #pragma once
 
+#include "ProcessQualified.h"
+#include <wtf/ArgumentCoder.h>
 #include <wtf/Hasher.h>
+#include <wtf/Markable.h>
 #include <wtf/URL.h>
 
 namespace WebCore {
 
 class Frame;
 class SecurityOrigin;
 
+enum class ShouldSkipOpaqueSecurityOriginCheck : bool { No, Yes };
+
 struct SecurityOriginData {
+    enum OpaqueOriginIdentifierType { };
+    using OpaqueOriginIdentifier = ProcessQualified<ObjectIdentifier<OpaqueOriginIdentifierType>>;
+    using MarkableOpaqueOriginIdentifier = Markable<OpaqueOriginIdentifier, OpaqueOriginIdentifier::MarkableTraits>;
+
     SecurityOriginData() = default;
     SecurityOriginData(const String& protocol, const String& host, std::optional<uint16_t> port)
         : protocol(protocol)
         , host(host)
         , port(port)
     {
         RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(!isHashTableDeletedValue());
+        if ((protocol.isEmpty() || protocol.isNull()) && (host.isEmpty() || host.isNull()))
+            *this = createOpaque();
```

Comment:
Why can't we just initialize `opaqueOriginIdentifier`?

Reply:
If protocol/host are emptyString(), then we should re-set them null, and calling `createOpaque` seemed cleaner. But, given my other response, I'll be changing this.